File: data-source/json_io.py
"""
Logical component: shared deterministic JSON writer for all Python data-source scripts.

Same input value -> byte-identical file output across runs and machines. Stays in
lockstep with scripts/json-io.mjs so Node and Python produce equivalent files
when both touch the same tree.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Callable, Iterable, Optional, Sequence


# Logical component: serialization defaults — 2-space indent, trailing newline, raw UTF-8.
DEFAULT_INDENT = 2


def _with_key_order(
    value: Any,
    key_order: Optional[Callable[[Sequence[str]], Iterable[str]]],
) -> Any:
    """Recursively reorder dict keys for deterministic output. Lists pass through."""
    if key_order is None:
        return value
    if isinstance(value, list):
        return [_with_key_order(v, key_order) for v in value]
    if isinstance(value, dict):
        ordered = list(key_order(list(value.keys())))
        return {k: _with_key_order(value[k], key_order) for k in ordered}
    return value


def write_json(
    path: Path | str,
    value: Any,
    *,
    key_order: Optional[Callable[[Sequence[str]], Iterable[str]]] = None,
) -> None:
    """
    Write JSON to disk deterministically: 2-space indent + trailing newline.

    Creates parent directories on demand. Pass `key_order` to enforce field
    ordering inside dicts (the callable receives the dict's existing keys and
    must return them in the desired order).
    """
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    ordered = _with_key_order(value, key_order)
    with p.open("w", encoding="utf-8") as f:
        json.dump(ordered, f, indent=DEFAULT_INDENT, ensure_ascii=False)
        f.write("\n")


def read_json(path: Path | str) -> Any:
    """Read and parse a JSON file. Companion to write_json."""
    with Path(path).open(encoding="utf-8") as f:
        return json.load(f)

File: scripts/json-io.mjs
// Logical component: shared deterministic JSON writer for all Node build scripts.
// Same input value → byte-identical file output across runs and machines.

import fs from 'node:fs';
import path from 'node:path';

// Logical component: serialization defaults — 2-space indent, trailing newline, no escaping of non-ASCII.
const DEFAULT_INDENT = 2;

/**
 * Reorder object keys for deterministic output. Recursive; arrays preserved as-is.
 * @param {unknown} value
 * @param {((keys: string[]) => string[]) | undefined} keyOrder
 * @returns {unknown}
 */
function withKeyOrder(value, keyOrder) {
  if (!keyOrder) return value;
  if (Array.isArray(value)) return value.map((v) => withKeyOrder(v, keyOrder));
  if (value === null || typeof value !== 'object') return value;
  const obj = /** @type {Record<string, unknown>} */ (value);
  const ordered = keyOrder(Object.keys(obj));
  /** @type {Record<string, unknown>} */
  const out = {};
  for (const k of ordered) out[k] = withKeyOrder(obj[k], keyOrder);
  return out;
}

/**
 * Write JSON to disk deterministically: 2-space indent + trailing newline.
 * Creates parent directories on demand.
 *
 * @param {string} filePath
 * @param {unknown} value
 * @param {{ keyOrder?: (keys: string[]) => string[] }} [options]
 */
export function writeJson(filePath, value, options = {}) {
  const dir = path.dirname(filePath);
  if (dir && !fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const ordered = withKeyOrder(value, options.keyOrder);
  fs.writeFileSync(filePath, `${JSON.stringify(ordered, null, DEFAULT_INDENT)}\n`);
}

/**
 * Read and parse a JSON file. Companion to `writeJson` so callers can avoid
 * re-importing `fs` for trivial reads.
 * @param {string} filePath
 * @returns {unknown}
 */
export function readJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}
